Check-In endpoint refactoring and cleanup

kruai opened the pull request:
Addresses https://issues.redhat.com/browse/RHCLOUD-10546

Glutexo commented:
@kruai I think we should raise 400 Bad Request if no canonical fact is provided.

Glutexo commented:
@kruai There are failing tests.

Glutexo commented:
I completely forgot about this. Sorry about that. 😞 May I ask to fix the failing tests @kruai and I‘ll try to not forgot again and review.

kruai commented:
@Glutexo Failing tests fixed!

Glutexo reviewed (changes requested):
Sorry about this. I didn’t realize that the Canonical Facts schema in the get Hosts query output is different from the input one. 😞

kruai commented on swagger/api.spec.yaml:
```diff
@@ -89,13 +89,10 @@ paths:
-                    host: Host details
         '404':
           description: Not Found.
+        '400':
+          description: Bad Request.
```
OK, so no changes are required here?

Glutexo replied:
No changes. 👍🏻

kruai commented on api/host.py:
```diff
@@ -376,6 +376,11 @@ def _build_paginated_host_tags_response(total, page, per_page, tags_list):
 def host_checkin(body):
     canonical_facts = deserialize_canonical_facts(body)
+    if not canonical_facts:
+        return flask_json_response(
+            {"detail": "No canonical facts provided.", "status": 400, "title": "Bad Request", "type": "about:blank"},
+            status=400,
+        )
```
I can change my responses to flask.abort, though I think changing the other ones in the file would be outside the scope of this PR.

kruai replied:
Updated.

kruai commented on swagger/api.spec.yaml:
```diff
@@ -672,47 +669,50 @@ components:
             An ID defined in /etc/insights-client/machine-id. This field is
             considered a canonical fact.
           type: string
+          nullable: true
```
Okay, I can change it so that it uses CanonicalFactsIn and CanonicalFactsOut. I wish there were a cleaner way to do that, though...

kruai commented on api/host.py:
```diff
@@ -376,6 +376,11 @@ def _build_paginated_host_tags_response(total, page, per_page, tags_list):
 @metrics.api_request_time.time()
 def host_checkin(body):
     canonical_facts = deserialize_canonical_facts(body)
+    if not canonical_facts:
```
OK, updated in the CanonicalFactsIn schema.

kruai commented:
@Glutexo I've separated the schema into CanonicalFactsIn and CanonicalFactsOut, and believe I've addressed all the issues you brought up. Can you please take another look?

Glutexo commented:
What happend to the tests? 😞

Glutexo reviewed (commented):
It looks like the suggested schema does not work as expected. I think I figured out a way. Trying this out now.

Glutexo commented:
@kruai I took a look at the schema and I think I found a way how to

- make Canonical Facts nullable on the output, but not on the input without repeating the whole definition,
- how to make on the Input at least one Canonical Fact required.

Please check my new commits to this pull request. What did I do:

- Extracted each Canonical Fact to its own component. This was not necessary, I just felt like it.
- Made all the Canonical Fact properties nullable. This needs to be the base state for further combination. Non-null values match both a _nullable schema_ and _non-nullable one_, null values match only a _nullable_ schema. Thus _allOf_ can be used add the _non-nullable_ restriction, but not to remove it.
- For the input, combined the output nullable Canonical Facts set with a set of schemas that only add `nullable: false` for each property. Unfortunately _nullable_ can’t be used just by itself, but always together with _type_. That’s because it’s just an OpenAPI alias for `type: [something, "null"]`.

kruai commented on swagger/api.spec.yaml:
```diff
@@ -89,13 +89,10 @@ paths:
-                    host: Host details
         '404':
           description: Not Found.
+        '400':
+          description: Bad Request.
```
OK. Considering my previous changes where I removed that particular flask.abort from the code, I think it makes sense to remove the `400` from here as well (since only deviating from the schema will result in a `400`).

Glutexo replied:
I agree. We however for some reason list 400 almost everywhere. 🤷🏻‍♂️

kruai commented on swagger/api.spec.yaml:
```diff
@@ -672,216 +669,227 @@ components:
+            $ref: "#/components/schemas/NonNullableArray"
+          external_id:
+            $ref: "#/components/schemas/NonNullableString"
+      - oneOf:
```
Shouldn't this be `anyOf`? `oneOf` tells it to accept one and only one item, whereas `anyOf` specifies one or more.

Glutexo replied:
Right!

kruai replied:
Replaced it and added some tests that validate these scenarios.

kruai reviewed (commented):
I think this needs 2 changes, one of which I'll do myself. The other change I want to run by you @Glutexo

Glutexo commented:
I pushed a few commits related to the checkin frequency parameter.

1. Added tests for the parameter and its validation.
2. Removed the default value as it’s not used by Connexion.
3. Removed the parameter and all the tests.

I think having the parameter there in the specification is only misleading. The value is not used at all and even if it’s not there, the clients can still use it and it will just be ignored.

If there is a reason to have it there, just revert the last commit and it will be back there with the tests I added. I can imagine that we want to make use of the validation not to break hosts already providing the value, once we implement its actual usage.

Glutexo reviewed (approved):
Thanks for your changes. Nice catch with the _oneOf_ combinator! 👍🏻
Please take a look at my last changes and revert the last commit if you don’t want the checkin frequency parameter to be removed.

kruai commented:
Your last changes look good, but we do want to keep the check-in frequency parameter. For now this is just supposed to update the host's updated date, but later on that value will be saved and used in different ways.

kruai commented:
@Glutexo how does this PR look now? Is it ready to be merged?

thearifismail reviewed (approved):
LGTM

Glutexo commented on swagger/api.spec.yaml:
```diff
@@ -672,216 +667,226 @@ components:
-          type: string
-          example: my.host.example.com
+          $ref: "#/components/schemas/IpAddresses"
+        Fqdn:
```
This breaks stuff!

```suggestion
      fqdn:
```

kruai replied:
Shit, how'd that get capitalized :notlikethis:

kruai replied:
Fix is here: https://github.com/RedHatInsights/insights-host-inventory/pull/826
